libr/util/udiff.c

```diff
@@ -38,6 +38,129 @@ R_API int r_diff_set_delta(RDiff *d, int delta) {
 	return 1;
 }
 
+typedef struct levrow {
+	ut32 *changes;
+	ut32 start, end;
+} Levrow;
+
+static void lev_matrix_free(Levrow *matrix, ut32 len) {
+	ut32 i;
+	for (i = 0; i < len; i++) {
+		if (matrix[i].changes) {
+			free (matrix[i].changes);
+		} else {
+			break;
+		}
+	}
+	free (matrix);
+}
+
+static inline int lev_byte_diff(RLevBuf *a, RLevBuf *b, ut32 ia, ut32 ib) {
+	if (a->mask) {
+		if ((a->buf[ia] & a->mask[ia]) == (b->buf[ib] & b->mask[ib])) {
+			return 0;
+		}
+	} else {
+		if (a->buf[ia] == b->buf[ib]) {
+			return 0;
+		}
+	}
+	return 1;
+}
+
+static inline Levrow *lev_row_init(Levrow *matrix, ut32 maxdst, ut32 rownum, ut32 buflen, ut32 delta) {
+	r_return_val_if_fail (matrix && !matrix[rownum].changes, false);
+	Levrow *row = matrix + rownum;
+
+	delta += rownum;
+	ut64 end = (ut64)delta + maxdst;
+	row->end = R_MIN (end, buflen);
+	row->start = delta <= maxdst? 0: delta - maxdst;
+
+	if ((row->changes = R_NEWS (ut32, row->end - row->start + 1)) == NULL) {
+		return NULL;
+	}
+	return row;
+}
+
+static inline ut32 lev_get_val(Levrow *row, ut32 i) {
+	if (i >= row->start && i <= row->end) {
+		return row->changes[i - row->start];
+	}
+	return UT32_MAX - 1; // -1 so a +1 with sub weight does not overflow
+}
+
+#define lev_set_next_op(arr, i, sz, op) \
+	if (i >= sz) { \
+		sz = sz * 2; \
+		if ((arr = reallocarray (arr, sizeof (RLevOp), sz)) == NULL) { \
+			return -1; \
+		} \
+	} \
+	arr[i++] = op;
+
+// obtains array of operations, in reverse order, to get from column to row of
+// matrix
+static st32 lev_parse_matrix(Levrow *matrix, ut32 len, bool invert, RLevOp **chgs) {
+	r_return_val_if_fail (len >= 2 && chgs && !*chgs, -1);
+
+	size_t size = len;
+	RLevOp *changes = R_NEWS (RLevOp, size);
+	if (!changes) {
+		return -1;
+	}
+
+	Levrow *row = matrix + len - 1;
+	Levrow *prev_row = row - 1;
+	RLevOp a = LEVADD;
+	RLevOp d = LEVDEL;
+
+	if (invert) {
+		a = LEVDEL;
+		d = LEVADD;
+	}
+
+	int j = row->end;
+	st32 insert = 0;
+	while (row != matrix) { // matrix[0] is not processed
+		ut32 sub = lev_get_val (prev_row, j - 1);
+		ut32 del = lev_get_val (prev_row, j);
+		ut32 add = lev_get_val (row, j - 1);
+
```

Comment:
you can realloc the array in here once. as long as every path below needs it, so you can avoid the use of the macro